security/backends/elasticsearch/writer.py

```diff
@@ -51,319 +56,291 @@ def get_querysets_by_batch(qs, batch):
-            logger.backend_logs.elasticsearch.update(
-                slug=logger.slug,
-                related_objects=self._get_related_object_keys(logger),
-                extra_data=logger.extra_data,
-                refresh=settings.ELASTICSEARCH_AUTO_REFRESH,
+    def create_or_update_index_from_logger(self, logger, is_last=False, **extra_data):
+        raise NotImplementedError
+
+
+class DirectElasticsearchDataWriter(BaseElasticsearchDataWriter):
+
+    def update_index(self, index, **updated_data):
+        try:
+            data = self._get_index_data(index)
+            data.update(updated_data)
+            index.update(
+                **self._update_data(data),
                 update_only_changed_fields=True,
                 retry_on_conflict=1,
-                **logger.data
+                refresh=settings.ELASTICSEARCH_AUTO_REFRESH,
             )
+        except ConflictError:
+            # conflict errors are ignored, celery task invocation was changed with another process
+            pass
 
-    def output_request_started(self, logger):
-        output_request_log = OutputRequestLog(
+    def _create_index_from_logger(self, logger, **extra_data):
+        index_class = get_log_model_from_logger_name(logger.name)
+        index = index_class(
             slug=logger.slug,
             release=logger.release,
             related_objects=self._get_related_object_keys(logger),
             extra_data=logger.extra_data,
-            state=RequestLogState.INCOMPLETE,
-            update_only_changed_fields=True,
-            parent_log=get_parent_log_key_or_none(logger),
-            **logger.data
+            parent_log=logger.parent_key,
+            **extra_data,
+            **self._update_data(logger.data)
         )
-        output_request_log.meta.id = logger.id
-        output_request_log.save()
-        logger.backend_logs.elasticsearch = output_request_log
+        index.meta.id = logger.id
+        index.save()
+        logger.backend_logs.elasticsearch_index = index
 
-    def output_request_finished(self, logger):
-        if 'elasticsearch' in logger.backend_logs:
-            logger.backend_logs.elasticsearch.update(
+    def _update_index_from_logger(self, logger, **extra_data):
+        self.update_index(logger.backend_logs.elasticsearch_index, **dict(
+            slug=logger.slug,
+            related_objects=self._get_related_object_keys(logger),
+            extra_data=logger.extra_data,
+            **extra_data,
+            **self._update_data(logger.data)
+        ))
+
+    def create_or_update_index_from_logger(self, logger, is_last=True, **extra_data):
+        if 'elasticsearch_index' in logger.backend_logs:
+            self._update_index_from_logger(logger, **extra_data)
+        else:
+            self._create_index_from_logger(logger, **extra_data)
+
+
+class LogstashElasticsearchDataWriter(BaseElasticsearchDataWriter):
+
+    MAX_ELASTICSEARCH_VERSION = 9999
```

Comment:
Optimistic conflict solution. Every write to the log increase version number. Elasticsearch check if new version is newest than currently stored. If version is lower or equal new log is not stored. For final log changes is used version 9999